lms/djangoapps/course_goals/migrations/0007_set_unsubscribe_token_default.py

```diff
@@ -0,0 +1,24 @@
+# Generated by Django 2.2.24 on 2021-08-20 19:18
+
+from django.db import migrations, models
+import uuid
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('course_goals', '0006_add_unsubscribe_token'),
+    ]
+
+    operations = [
+        migrations.AlterField(
+            model_name='coursegoal',
+            name='unsubscribe_token',
+            field=models.UUIDField(blank=True, default=uuid.uuid4, editable=False, help_text='Used to validate unsubscribe requests without requiring a login', null=True, unique=True),
+        ),
+        migrations.AlterField(
+            model_name='historicalcoursegoal',
+            name='unsubscribe_token',
+            field=models.UUIDField(blank=True, db_index=True, default=uuid.uuid4, editable=False, help_text='Used to validate unsubscribe requests without requiring a login', null=True),
+        ),
```

Comment:
Should we exempt this field from the historical model? I wonder if there is value (or concern?) in saving

Reply:
It's a readonly field, so it's not expected to change - but if it does change from null to not, or some other way, I guess good to know about it and have a record? I'm not stressed about the extra size, since it won't cause much churn itself.

Maybe you are worried about more places where this "password token" is kept? But the existing plan is to just have these uuids sit in db long term. Because (A) I don't think there's a huge risk if someone sniffs this token in emails or whatever - can just unsubscribe someone and (B) I want even old emails to have valid unsubscribe links. So I don't _think_ I'm worried about extra risk in this token leaking?

Reply:
The only worry is if this gets taken as a patten that can be used for password reset or something

Reply:
Yeah, hopefully they would be security thinking enough to not replicate this logic for something more sensitive. But I see your point MT. I'm fine with you leaving it.

And yes, my concern was more on the "hey, we know this link can be used to unsubscribe learners, maybe it shouldn't be so long lasting or preserved in case it was leaked", but I think you're right about risk is low and enabling old emails is good. Lot easier overhead too without needing to worry about clearing/invalidating keys.

Reply:
I'm hoping the comments on both sides of this (model and endpoint) will cause anyone that is copying and pasting to at least consider the security side of things / not add more stuff or PII to the endpoint.

But I'm open to thoughts on sliding the "convenience vs security" slider more towards security. As a thought experiment, that would involve... Maybe adding an expiration datetime field and then having each email campaign job refresh the token / reset the deadline, in bulk? Where the cost would be code complexity and old email unsubscribe links not working. My gut isn't enthused about the trade-off, but it's not horrible. I could write it if you folks are vibing it.